stdlib/asm/collections/smt_new.masm

```diff
@@ -0,0 +1,161 @@
+const.DEPTH=64
+
+#! Input:  [V, K, R, ...]
+#! Output: [V_old, R_new, ...]
+export.set
+    adv.push_smtpeek adv_push.4
+    # => [V_old, V, K, R]
+
+    padw eqw
+    # => [V_old == ZERO, ZERO, V_old, V, K, R]
+
+    if.true
+        # empty leaf
+
+        dropw dropw
+        #=> [V, K, R]
+
+        # Update advice map with new leaf preimage
+        # adv_map[hash(K, V)] = [K, V]
+        adv.insert_hdword
+        #=> [V, K, R]
+
+        # Prepare stack for `mtree_set`
+        movupw.2 dup.11 push.DEPTH
+        #=> [depth, K[3], R, V, K]
+
+        # Insert node in Merkle store
+        mtree_set
+        #=> [V_in_leaf, R_new, K]
+
+        # Check that V_in_leaf is indeed empty
+        padw assert_eqw
+        #=> [R_new, K]
+
+        # Prepare stack for return
+        swapw dropw padw
+        #=> [ZERO, R_new]
+    else
+        # Single or Multiple leaf
+
+        dropw
+        # => [V_old, V, K, R]
+
+        # Retrieve leaf pre-image on advice stack, and push leaf size on stack
+        adv.push_mapval adv_push.1
```

Comment:
I believe we need to use `adv.push_mapvaln` here to also get the length of the list onto the advice stack.